.github/workflows/ci.yml

```diff
@@ -17,8 +17,15 @@ jobs:
       matrix:
         distro: [centos.7, leap.15, ubuntu.20.04]
     steps:
-    - uses: actions/checkout@v2
+    - name: Checkout code
+      uses: actions/checkout@v2
       with:
         submodules: true
-    - name: Build Docker image
-      run: docker build . -f utils/docker/Dockerfile.${{ matrix.distro }} -t ${{ matrix.distro }}/daos:${{ github.sha }}
+    - name: Build ${{ matrix.distro }} Docker image with gcc
+      run: docker build . -f utils/docker/Dockerfile.${{ matrix.distro }}
+                          --build-arg COMPILER=gcc
+                          -t ${{ matrix.distro }}/gcc/daos:${{ github.sha }}
+    - name: Build ${{ matrix.distro }} Docker image with clang
+      run: docker build . -f utils/docker/Dockerfile.${{ matrix.distro }}
+                          --build-arg COMPILER=clang
+                          -t ${{ matrix.distro }}/clang/daos:${{ github.sha }}
```

Comment:
While I agree that doing this serially as you have done can make the second build quicker because it re-uses the Dockerfile, if instead you used multiple axes such as:

```
      matrix:
￼        distro: [centos.7, leap.15, ubuntu.20.04]
￼        compiler: [gcc, clang]
￼```

you would have builds for both compilers on each distro all running in parallel, which should ultimately finish more quickly than even re-using the existing Dockerfile in a serial build fashion.

The other problem with serial builds is that one has to dig into each distro to figure out which compiler build broke rather than having the particular distro/compiler combination showing quite obviously as broken.

Reply:
That was my original version here (https://github.com/daos-stack/daos/pull/4211/commits/f7515a54d7ecc0f09184a8964a8809b96ea36e15#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR19)
I can resume to this version if that's preferred.

Reply:
> I can resume to this version if that's preferred.

I think so.  I think it's very worthwhile making the failure as specific and obvious as possible.

Really/ideally, GitHub should be caching docker images so Docker image creation should be almost free -- the way it is in our Jenkins CI pipeline (assuming the Dockerfile has not been modified).

If they are not doing this already, we can hope that they see the light and start doing this.  Or maybe for them. storage is more expensive than CPU cycles/build execution time/etc.